frontend/src/features/verifiable-fields/VerifiableFieldProvider.tsx

```diff
@@ -0,0 +1,148 @@
+import { useCallback, useState } from 'react'
+import {
+  ControllerRenderProps,
+  useFormContext,
+  useWatch,
+} from 'react-hook-form'
+
+import { FormFieldWithId } from '~shared/types/field'
+import { isMobilePhoneNumber } from '~shared/utils/phone-num-validation'
+
+import { BaseFieldProps } from '~templates/Field/FieldContainer'
+
+import {
+  VerifiableFieldBase,
+  VerifiableFieldInput,
+  VerifiableFieldSchema,
+} from './types'
+import { VerifiableFieldContext } from './VerifiableFieldContext'
+
+export interface VerifiableFieldProps extends BaseFieldProps {
+  schema: VerifiableFieldSchema<FormFieldWithId<VerifiableFieldBase>>
+}
+
+export interface VerifiableFieldProviderProps extends VerifiableFieldProps {
+  children: React.ReactNode
+}
+
+export const VerifiableFieldProvider = ({
+  schema,
+  children,
+}: VerifiableFieldProviderProps): JSX.Element => {
+  const [isVfnBoxOpen, setIsVfnBoxOpen] = useState(false)
+
+  const { control, setError, getValues, setValue, setFocus, clearErrors } =
+    useFormContext()
+  const currentSignature: VerifiableFieldInput = useWatch({
+    name: `${schema._id}.signature`,
+    control,
+  })
+
+  /**
+   * Keeps track of already completed numbers to signatures
+   */
+  const [mapNumberToSignature, setMapNumberToSignature] = useState<
+    Record<string, string>
+  >({})
+
+  const handleInputChange = useCallback(
+    (onChange: ControllerRenderProps['onChange']) => (value?: string) => {
+      // Prevent action when multiple onChange is called with the same value
+      // This can happen when input is blurred, since onChange is also called by
+      // react-hook-form when that happens.
+      if (getValues(schema._id)?.value === value) return
+
+      // If the input changes then the user needs to request re-verification,
+      // and the vfn box should be hidden.
+      if (isVfnBoxOpen) {
+        setIsVfnBoxOpen(false)
+      }
+      if (!value) {
+        return onChange({ value })
+      }
+
+      // If the current input corresponds to a signature, update the signature
+      // value of this field.
+      const signature = mapNumberToSignature[value ?? '']
```

Comment:
yes; should be undefined. Typescript has a "quirk" in that return type of index access will be assumed to be the declared type.

TS actually has a new flag to prevent that, but it introduces a lot of noise in the code and thus we probably don't want to enabled that. See [docs](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-beta/#no-unchecked-indexed-access)